File: pull-up-backend/src/main/java/me/graceyan/pull_up_backend/model/Event.java
package me.graceyan.pull_up_backend.model;

import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.bson.types.ObjectId;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.data.mongodb.core.mapping.DocumentReference;

import java.time.DayOfWeek;
import java.time.Instant;
import java.time.LocalDate;
import java.time.LocalTime;
import java.util.List;

@Document(collection = "events")
@Data
@AllArgsConstructor
@NoArgsConstructor
public class Event {
    @Id
    private ObjectId id;

    @Indexed(unique = true)
    private String publicId;

    @Indexed(unique = true)
    private String urlAlias; // publicId by default

    private String name;

    private String type; // Weekly, DayTime, Day Only

    private String timezone; // Default timezone

    private List<DayOfWeek> excludeDaysOfWeek;

    private LocalTime startTime;

    private LocalTime endTime;

    private LocalDate startDate; // "2025-08-15"

    private LocalDate endDate;

    @CreatedDate
    private Instant createdAt;

    @DocumentReference
    private List<User> userIds;

    public Event(
            String publicId, String urlAlias, String name, String type,
            String timezone, List<DayOfWeek> excludeDaysOfWeek, LocalTime startTime, LocalTime endTime,
            LocalDate startDate, LocalDate endDate)
    {
        this.publicId = publicId;
        this.urlAlias = urlAlias;
        this.name = name;
        this.type = type;
        this.timezone = timezone;
        this.excludeDaysOfWeek = excludeDaysOfWeek;
        this.startTime = startTime;
        this.endTime = endTime;
        this.startDate = startDate;
        this.endDate = endDate;
    }
}

File: frontend/src/lib/event.ts
import { dateToTimeString, formatDate, type DayOfWeek } from "./calendar";
import type { User } from "./user";

export type EventType = "day" | "dayTime" | "weekdayTime";

export type Event = {
  id: string;
  publicId: string;
  urlAlias: string;
  name: string;
  type: EventType;
  timezone: string;
  excludeDaysOfWeek: DayOfWeek[];
  startTime: Date;
  endTime: Date;
  startDate: Date;
  endDate: Date;
  createdAt: Date;
  userIds: User[];
};

export type EventRaw = {
  id: string;
  publicId: string;
  urlAlias: string;
  name: string;
  type: EventType;
  timezone: string;
  excludeDaysOfWeek: DayOfWeek[];
  startTime: string;
  endTime: string;
  startDate: string;
  endDate: string;
  createdAt: string;
  userIds: User[];
};

export type EventPayload = {
  urlAlias?: string;
  name: string;
  type: EventType;
  timezone: string;
  excludeDaysOfWeek: DayOfWeek[] | null;
  startTime: string | null;
  endTime: string | null;
  startDate: string | null;
  endDate: string | null;
};

export type EventFormData = {
  name: string;
  urlAlias?: string;
  startDate?: Date;
  endDate?: Date;
  startTime?: Date;
  endTime?: Date;
};

export function convertToEventPayload(
  eventType: EventType,
  formData: EventFormData
) {
  return {
    ...formData,
    startDate: formData.startDate ? formatDate(formData.startDate) : null,
    endDate: formData.endDate ? formatDate(formData.endDate) : null,
    startTime: formData.startTime ? dateToTimeString(formData.startTime) : null,
    endTime: formData.endTime ? dateToTimeString(formData.endTime) : null,
    type: eventType,
    timezone: Intl.DateTimeFormat().resolvedOptions().timeZone,
    excludeDaysOfWeek: [], // TODO
  };
}
